libosmscout-client-qt/src/osmscout/DBJob.cpp

```diff
@@ -0,0 +1,193 @@
+{
+  Close();
+}
+
+void DBJob::Run(QList<DBInstanceRef> &databases, QReadLocker *locker)
+{
+  if (thread!=QThread::currentThread()){
+    qWarning() << "Run from non Job thread";
+  }
+  this->databases=databases;
+  this->locker=locker;
+}
+
+void DBJob::Close()
+{
+  if (locker==NULL){
+    return;
+  }
+  if (thread!=QThread::currentThread()){
+    qWarning() << "Closing from non Job thread";
+  }
+  delete locker;
+  databases.clear();
+}
+
+DBLoadJob::DBLoadJob(osmscout::MercatorProjection lookupProjection,
+                     unsigned long maximumAreaLevel,
+                     bool lowZoomOptimization):
+  DBJob(),
+  breaker(std::make_shared<QBreaker>()),
+  lookupProjection(lookupProjection)
+{
+  searchParameter.SetMaximumAreaLevel(maximumAreaLevel);
+  searchParameter.SetUseMultithreading(true);
+  searchParameter.SetUseLowZoomOptimization(lowZoomOptimization);
+  searchParameter.SetBreaker(breaker);
+  
+  connect(this,SIGNAL(tileStateChanged(QString,const osmscout::TileRef)),
+          this,SLOT(onTileStateChagned(QString,const osmscout::TileRef)),
+          Qt::QueuedConnection);
+}
+
+DBLoadJob::~DBLoadJob()
+{
+}
+
+void DBLoadJob::Run(QList<DBInstanceRef> &databases, QReadLocker *locker)
+{
+  osmscout::GeoBox lookupBox;
+  lookupProjection.GetDimensions(lookupBox);
+  QList<DBInstanceRef> relevantDatabases;
+  for (auto &db:databases){
+    if (!db->database->IsOpen() || (!db->styleConfig)) {
+      qDebug() << "Database is not ready" << db->path;
+      continue;
+    }
+    osmscout::GeoBox dbBox;
+    db->database->GetBoundingBox(dbBox);
+    if (!dbBox.Intersects(lookupBox)){
+      qDebug() << "Skip database" << db->path;
+      continue;
+    }
+    relevantDatabases << db;
+  }
+
+  DBJob::Run(relevantDatabases,locker);
+  for (auto &db:relevantDatabases){
+    std::list<osmscout::TileRef> tiles;
+    db->mapService->LookupTiles(lookupProjection,tiles);
+
+    QString path=db->path;
+    osmscout::MapService::TileStateCallback callback=[this,path](const osmscout::TileRef& tile) {
+      emit tileStateChanged(path,tile);
+    };
+    
+    osmscout::MapService::CallbackId callbackId=db->mapService->RegisterTileStateCallback(callback);
+    callbacks[db->path]=callbackId;
+    loadedTiles[db->path]=QMap<osmscout::TileId,osmscout::TileRef>();
+    QMap<osmscout::TileId,osmscout::TileRef> tileMap;
+    for (auto &tile:tiles){
+      tileMap[tile->GetId()]=tile;
+    }
+    loadingTiles[db->path]=tileMap;
+
+    // load tiles asynchronous
+    db->mapService->LoadMissingTileDataAsync(searchParameter,
+                                             *db->styleConfig,
+                                             tiles);
+
+    // process already completed tiles (state callback is not called in such case)
+    for (auto &tile:tiles){
+      if (tile->IsComplete()){
+        emit tileStateChanged(path,tile);
+      }
+    }
+
+  }
+}
+
+void DBLoadJob::onTileStateChagned(QString dbPath,const osmscout::TileRef tile)
```

Comment:
const QString& for dbPath to optimize? I also use const& for XXXRef classes, though that usage could be discussed.